src/backend/distributed/transaction/backend_data.c

```diff
@@ -1278,3 +1291,214 @@ DecrementExternalClientBackendCounter(void)
 {
 	pg_atomic_sub_fetch_u32(&backendManagementShmemData->externalClientBackendCounter, 1);
 }
+
+
+/*
+ * GetEpochTimeMs returns the epoch value in milliseconds.
+ */
+uint64
+GetEpochTimeMs(void)
+{
+	struct timeval tp;
+
+	gettimeofday(&tp, NULL);
+
+	uint64 result = (uint64) (tp.tv_sec) * 1000;
+	result = result + (uint64) (tp.tv_usec) / 1000;
+	return result;
+}
+
+
+/*
+ * citus_get_cluster_clock() is an UDF that returns a monotonically increasing
+ * logical clock. Clock guarantees to never go back in value after restarts, and
+ * makes best attempt to keep the value close to unix epoch time in milliseconds.
+ */
+Datum
+citus_get_cluster_clock(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	uint64 nextClusterClockValue = GetNextClusterClock();
+
+	PG_RETURN_UINT64(nextClusterClockValue);
+}
+
+
+/*
+ * GetNextClusterClock implements the internal guts of the UDF citus_get_cluster_clock()
+ */
+uint64
+GetNextClusterClock(void)
+{
+	uint64 epochValue = GetEpochTimeMs();
+
+	SpinLockAcquire(&backendManagementShmemData->clockMutex);
+
+	/* Check if the clock is adjusted after the boot */
+	if (backendManagementShmemData->clusterClockValue == 0)
+	{
+		SpinLockRelease(&backendManagementShmemData->clockMutex);
+		ereport(ERROR, (errmsg("backend never adjusted the clock, please retry")));
+	}
+
+	uint64 nextClusterClockValue = backendManagementShmemData->clusterClockValue + 1;
+
+	/*
+	 * Incrementing by 1 ensures uniqueness as the clock precision can be
+	 * less than a millisecond.
+	 */
+	epochValue = epochValue + 1;
```

Comment:
I don't understand why we would need to increment the epoch here.
Wouldn't the uniqueness be guaranteed by the virtue of the comparison below? Where we take `MAX(epoch, nextClusterClockValue)`?

If we would read the same epoch multiple times it guarantees to keep incrementing because of the increment right above here, together with the `MAX(...)`.

Reply:
You are correct, I was trying to protect for repeated calls within one millisecond, but as we save the previous value and increment it, it's guaranteed to have an increasing value. Removed it now.